{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "88437e91",
   "metadata": {},
   "source": [
    "## AnyVar VCF processing and annotation\n",
    "\n",
    "### Setup\n",
    "\n",
    "First, we'll initialize AnyVar (we already have some required services running in the background) and the VCF registrar object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "04768838",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krs021/anyvar/src/anyvar/__init__.py:5: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "/Users/krs021/anyvar/.venv/lib/python3.11/site-packages/pydantic/_internal/_config.py:373: UserWarning: Valid config keys have changed in V2:\n",
      "* 'schema_extra' has been renamed to 'json_schema_extra'\n",
      "  warnings.warn(message, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from timeit import default_timer as timer\n",
    "\n",
    "from anyvar.anyvar import AnyVar, create_storage, create_translator\n",
    "from anyvar.extras.vcf import VcfRegistrar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc06a554",
   "metadata": {},
   "source": [
    "Running the next command assumes that you have a local seqrepo at usr/local/share/seqrepo/2024-12-20. You can also change it to the RestAPI seqrepo URI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3350e9e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import environ\n",
    "environ[\"SEQREPO_DATAPROXY_URI\"] = \"seqrepo+file:///usr/local/share/seqrepo/2024-12-20\"\n",
    "\n",
    "av = AnyVar(\n",
    "    create_translator(),\n",
    "    create_storage(\"postgresql://postgres:postgres@localhost:5432/anyvar\"),\n",
    ")\n",
    "vcf_registrar = VcfRegistrar(data_proxy=av.translator.dp, av=av)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52a0eda7",
   "metadata": {},
   "source": [
    "### Input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d33e65e8",
   "metadata": {},
   "source": [
    "We have a sample file `demo-input.vcf`, with about 1,000 rows comprised of simple SNPs and indels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fec818f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    1045 ./demo-input.vcf\n"
     ]
    }
   ],
   "source": [
    "!wc -l ./demo-input.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be26b480",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b]10;?\u0007\u001b]11;?\u0007\u001b[c\u001b[38;5;246m───────┬────────────────────────────────────────────────────────────────────────\u001b[0m\n",
      "       \u001b[38;5;246m│ \u001b[0mFile: \u001b[1m./demo-input.vcf\u001b[0m\n",
      "\u001b[38;5;246m───────┼────────────────────────────────────────────────────────────────────────\u001b[0m\n",
      "\u001b[38;5;246m───────┴────────────────────────────────────────────────────────────────────────\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!bat --line-range=0:100 ./demo-input.vcf  # for example"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ef7fdb",
   "metadata": {},
   "source": [
    "### Ingestion and annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa063536",
   "metadata": {},
   "source": [
    "We'll run the `annotate()` method and track wall clock time:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "991a0844",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processed all VCF rows in 3.655102666991297 seconds\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "start = timer()\n",
    "vcf_registrar.annotate(Path(\"./demo-input.vcf\"), vcf_out=Path(\"./demo-output.vcf\"))\n",
    "end = timer()\n",
    "print(f\"processed all VCF rows in {end - start} seconds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f41d379",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ac158315",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Between references and alternates, this registers 1872 alleles.\n"
     ]
    }
   ],
   "source": [
    "allele_count = av.object_store.get_variation_count(\"all\")\n",
    "print(f\"Between references and alternates, this registers {allele_count} alleles.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc5fac0a",
   "metadata": {},
   "source": [
    "### Output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "634346c1",
   "metadata": {},
   "source": [
    "This process adds VRS allele IDs to the VCF's INFO field:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ee3869ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b]10;?\u0007\u001b]11;?\u0007\u001b[c\u001b[38;5;246m───────┬────────────────────────────────────────────────────────────────────────\u001b[0m\n",
      "       \u001b[38;5;246m│ \u001b[0mFile: \u001b[1m./out.vcf\u001b[0m   <EMPTY>\n",
      "\u001b[38;5;246m───────┴────────────────────────────────────────────────────────────────────────\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!bat --line-range=0:100 ./out.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed38fdd",
   "metadata": {},
   "source": [
    "We can dereference those IDs to retrieve the complete VRS allele:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc8e42a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "av.get_object(\"ga4gh:VA.tT2-U2WwLDM0r77vQwCu3amz8fCkuVw_\", True).as_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83632a74",
   "metadata": {},
   "source": [
    "### Search\n",
    "\n",
    "Currently, we support basic genomic region searches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e421a7c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr4 = av.translator.get_sequence_id(\"NCBI:NC_000004.12\")\n",
    "av.object_store.search_variations(chr4, 400000, 500000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c68bb76a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
